[#115] Add ExceptionHandler interface that users can customize exception handler

Alencion opened the pull request:
## Description
I try to implement customExceptionHandler feature. 

like this. 
### 1. definition Exception Handler
``` java
public class GlobalExceptionHandler implements DssExceptionHandler {
     // global exception handle method
    @ExceptionHandler(exception = NullPointerException.class)
    public YourServiceResponse globalExceptionHandleMethod(){
        YourServiceResponse yourServiceResponse = new YourServiceResponse();
        yourServiceResponse.setResult("Handle Exception on globalExceptionHandlerMethod");
        return yourServiceResponse;
    }

    // service exception handler method
    @ServiceExceptionHandler(service = MyService.class, exception = NullPointerException.class)
    public MyServiceResponse serviceExceptionHandleMethod(MyServiceRequest request){
        MyServiceResponse myServiceResponse = new MyServiceResponse();
        myServiceResponse.setResult("Occur NullPointer Exception in MyService." + " request: " + request.getName());
        return myServiceResponse;
    }
}
```

### 2. add exception handler 
```java
publicclass Application {

    @Test
    public static void main() throws InterruptedException {
        final DssRestServer dssRestServer = new DssRestServer("127.0.0.1", 8181);
        dssRestServer
                .addDssService(new MyService())
                .addDssService(new YourService())
                .addExceptionHandler(new GlobalExceptionHandler());

        dssRestServer.start();
    }
}
```

### 3. handling exception
global
![image](https://user-images.githubusercontent.com/15667978/91011722-c9aaa580-e61f-11ea-8804-67ae4da133be.png)

service
![image](https://user-images.githubusercontent.com/15667978/91011752-d7f8c180-e61f-11ea-9d78-1ad1e7b6c257.png)


---
##  What i want a review point

1. method, variable name etc.
2. Whether I created the java file in the right package or not.
3. if DssRestRequestMappingException occur, The exception handle method with request parameter does't execute. What do you think this situation?
4. anything else.

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- add ExceptionHandler annotations
- add DssRestExceptionHandler
- add DssRestExceptionHandlerActor

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- update exceptionHandleMethod to receive parameter

ztkmkoo commented on dss-core/src/main/java/io/github/ztkmkoo/dss/core/message/rest/DssRestExceptionHandlerCommandRequest.java:
```diff
@@ -0,0 +1,21 @@
+public class DssRestExceptionHandlerCommandRequest implements DssRestExceptionHandlerCommand {
+    private static final long serialVersionUID = 5752260184738368572L;
+
+    private final DssRestActorService service;
```
It looks transient(not serializable)?
If yes, please add the keyword

Alencion replied:
Thank you for review! 

I couldn't understand your opinion. Could you explain more detail?

ztkmkoo replied:
transient is a variables modifier used in serialization. At the time of serialization, if we don’t want to save value of a particular variable in a file, then we use transient keyword. When JVM comes across transient keyword, it ignores original value of the variable and save default value of that variable data type.


The code may be
``` java
public class DssRestExceptionHandlerCommandRequest implements DssRestExceptionHandlerCommand {
    private static final long serialVersionUID = 5752260184738368572L;

    private transient final DssRestActorService service;
```

ztkmkoo commented on dss-server/src/main/java/io/github/ztkmkoo/dss/server/DssServer.java:
```diff
@@ -38,4 +39,6 @@
      */
     DssServer<S> addDssService(S service);
+
+    DssServer<S> addExceptionHandler(DssExceptionHandler exceptionHandler);
```
Could you like to add the document of this method?

Alencion replied:
Ok. I will add document.

ztkmkoo commented on dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestExceptionHandlerActor.java:
```diff
@@ -0,0 +1,124 @@
+            Constructor<?> declaredConstructor = method.getDeclaringClass().getDeclaredConstructor();
+            declaredConstructor.setAccessible(true);
+
+            Object exceptionHandler = declaredConstructor.newInstance();
```
Do i have to generate new exception handler instance every exception handling?

Alencion replied:
I think it just need one. I didn't think about this. Thank you.

I try to code an instance of exceptionHandler to create once.

ztkmkoo commented on dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestExceptionHandlerActor.java:
```diff
@@ -0,0 +1,124 @@
+                return (DssRestServiceResponse) response;
+            }
+        } catch (InstantiationException | InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
+            e.printStackTrace();
```
Please use logger instead of printStackTrace

Alencion replied:
Ok. I'll try logger.

ztkmkoo commented on dss-core/src/main/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestExceptionHandlerActor.java:
```diff
@@ -0,0 +1,124 @@
+            Object response;
+
+            if (method.getParameterCount() == 1){
+                Method makeRequest = AbstractDssRestActorService.class.getDeclaredMethod("makeRequest", DssRestServiceActorCommandRequest.class);
+                makeRequest.setAccessible(true);
+
+                DssRestServiceRequest request = (DssRestServiceRequest) makeRequest.invoke(commandRequest.getService(), commandRequest.getRequest());
```
Is it the best way to get the service request?

Alencion replied:
if this method's access modifier may be changed to public,
it seems possible like the next code.

```java
DssRestServiceRequest request = commandRequest.getService().makeRequest(commandRequest.getRequest());

response = method.invoke(exceptionHandler, request.getBody());
```

ztkmkoo replied:
Ok~
It looks better.

ztkmkoo commented:
@Alencion 
Is it necessary using reflection in exception handling actor?
I want to talk bout with you~~

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- delete e.printStackTrace()

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- modify exceptionHandlerMap using ExceptionHandleMethod

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- add transient in DssRestExceptionHandlerCommandRequest

ztkmkoo commented:
> 
> 
> [![Coverage Status](https://camo.githubusercontent.com/71284c450f0a16d36c234fa905f79aa2aabcc666/68747470733a2f2f636f766572616c6c732e696f2f6275696c64732f33333037323632332f6261646765)](https://coveralls.io/builds/33072623)
> 
> Coverage decreased (-6.9%) to 76.183% when pulling **[07884d9](https://github.com/ztkmkoo/dss/commit/07884d90214e7feb2c87a7dbbfb1ec217d29efe1) on Alencion:feature/115_exception_handler** into **[87398b2](https://github.com/ztkmkoo/dss/commit/87398b2af825fd16eea8603523326ae3f9abe657) on ztkmkoo:develop**.

@Alencion 
Please success the coveralls coverage~

ztkmkoo commented on dss-core/src/main/java/io/github/ztkmkoo/dss/core/exception/handler/DssRestExceptionHandler.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.Map;
+import java.util.Objects;
+
+public class DssRestExceptionHandler {
```
Hi,
The name of the class(`DssRestExceptionHandler`) looks implementing the DssExceptionHandler interface.

How bout change name to DssRestExceptionHanlderResolver or else?

Alencion replied:
I agree with that. The DssRestExceptionHandlerResolver is good name.

Naming is the hardest thing in the world. 👍

ztkmkoo commented on dss-core/src/main/java/io/github/ztkmkoo/dss/core/exception/handler/DssRestExceptionHandler.java:
```diff
@@ -0,0 +1,70 @@
+
+    private ExceptionHandleMethod setExceptionHandleMethod(DssExceptionHandler dssExceptionHandler, Method method){
+        return request -> {
+            try {
+                return (DssRestServiceResponse) method.invoke(dssExceptionHandler, request.getBody());
+            } catch (Exception e) {
+                logger.error("exception handling error: ", e);
+            }
+            return null;
```
Why return null?
Should return UNKOWN ERROR OR SERVER INTERNAL ERROR i think.

Alencion replied:
My mistake. I don't think about that.

ztkmkoo replied:
We(Me and @Alencion) decide to return null here.

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- add document addExceptionHandler Method

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- update exceptionHandleMethod catch statement

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- rename DssRestExceptionHandler -> DssRestExceptionHandlerResolver

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- rename exceptionHandler -> exceptionHandlerActor in DssRestMasterActor

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- modify global exception handler method params type
  Serializable -> DssRestServiceActorCommandRequest

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- modfiy code style

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- add exception handle method accessible

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- add test

ztkmkoo commented on dss-server/src/test/java/io/github/ztkmkoo/dss/server/rest/DssRestExceptionHandlerTest.java:
```diff
@@ -0,0 +1,128 @@
+class DssRestExceptionHandlerTest {
+
+    @Test
+    void test() throws InterruptedException {
+        final DssRestServer dssRestServer = new DssRestServer("127.0.0.1", 8181);
+        dssRestServer
+                .addDssService(new MyService())
+                .addDssService(new YourService())
+                .addExceptionHandler(new GlobalExceptionHandler());
+
+        stopDssRestServerAfterActivated(dssRestServer, 10 ,15);
+
+        dssRestServer.start();
+
+        assertTrue(dssRestServer.isShutdown());
+    }
```
@Alencion 
This test looks a little bit strange. Could you tell me the detail?

Alencion replied:
I want to test that the server operates well after addExceptionHandler is added.

ztkmkoo replied:
Yap, i see.

But the Test class name is `DssRestExceptionHandlerTest.java`. 
I was mistaken that you test the `ExceptionHandler.java` interface.
You already test it in `DssRestServerTest.java` why test here?

ztkmkoo commented on dss-core/src/test/java/io/github/ztkmkoo/dss/core/actor/rest/DssRestExceptionHandlerActorTest.java:
```diff
@@ -0,0 +1,65 @@
+                .exception(new Exception())
+                .build());
+
+        probe.expectMessageClass(DssRestChannelHandlerCommandResponse.class);
```
Could you test more bout the response not only the expected class.
I think we can expect the status code in the response at leat.

Alencion replied:
Ok. I don't think about it.

I will add status code and channelid assertEquals.
```java
assertEquals(HttpResponseStatus.BAD_REQUEST.code(), response.getStatus().intValue());
assertEquals("test", response.getChannelId());
```

ztkmkoo commented on dss-core/src/test/java/io/github/ztkmkoo/dss/core/exception/handler/DssRestExceptionHandlerResolverTest.java:
```diff
@@ -0,0 +1,54 @@
+            }
+        });
+
+        assertFalse(dssRestExceptionHandlerResolver.getExceptionHandlerMap().isEmpty());
```
Can we expect the key contains in the map? Test more like
``` java
assertTrue(dssRestExceptionHandlerResolver.getExceptionHandlerMap().contains(TestService.class))
```

Alencion replied:
I agree with that! I think that's a better test.

ztkmkoo replied:
Thank you so much

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- add test

ztkmkoo reviewed (approved):
Thank you so much.

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- add test and modify code style

Alencion committed:
[#115] Add ExceptionHandler interface that users can customize exception
handler

- add DssTestExceptionHandlerActor test

Doyuni commented on dss-core/src/main/java/io/github/ztkmkoo/dss/core/exception/handler/DssRestExceptionHandlerResolver.java:
```diff
@@ -0,0 +1,74 @@
+            if (method.isAnnotationPresent(ExceptionHandler.class)) {
+                ExceptionHandler annotation = method.getAnnotation(ExceptionHandler.class);
+
+                for (Class<? extends Exception> exception : annotation.exception()) {
+                    globalExceptionHandlerMap.put(exception, setExceptionHandleMethod(dssExceptionHandler, method));
+                }
```
I'd like to share what my concern is. This is same case on lines 49~51 too.
If `setExceptionHandlerMethod()` return null, `cotainMethod()` method of `DssRestExceptionHandlerActor` class may be return true.
For example)
```java
Map<String, Integer> map = new HashMap<>();
map.put("hello", null);
System.out.println(map.containsKey("hello")); // true
```

So, I recommend that this code will be changed like below.
```java
for (Class<? extends Exception> exception : annotation.exception()) {
    ExceptionHandleMethod exceptionHandleMethod = setExceptionHandleMethod(dssExceptionHandler, method);
    
    if (Objects.nonNull(exceptionHandleMethod)) {
            globalExceptionHandlerMap.put(exception, exceptionHandleMethod);
    }
 }
```

Alencion replied:
Thank you for review!! @Doyuni 

I think this method can't return null.

The reasons are as follows:
setExceptionHandleMethod(dssExceptionHandler, method); is return new ExceptionHandleMethod instance. 
```java
    private ExceptionHandleMethod setExceptionHandleMethod(DssExceptionHandler dssExceptionHandler, Method method) {
        return request -> {
            try {
                method.setAccessible(true);
                return (DssRestServiceResponse) method.invoke(dssExceptionHandler, request);
            } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
                logger.error("exception handling method invoke error: ", e);
            } catch (Exception e) {
                logger.error("exception handling error: ", e);
            }
            return null;
        };
    }

    // it is same as the previous code.
    private ExceptionHandleMethod setExceptionHandleMethod(DssExceptionHandler dssExceptionHandler, Method method) {
        return new ExceptionHandleMethod() {
            @Override
            public DssRestServiceResponse handlingException(DssRestServiceActorCommandRequest request) {
                try {
                    method.setAccessible(true);
                    return (DssRestServiceResponse) method.invoke(dssExceptionHandler, request);
                } catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
                    logger.error("exception handling method invoke error: ", e);
                } catch (Exception e) {
                    logger.error("exception handling error: ", e);
                }
                return null;
            }
        };
    }
```

So, i think it don't need null checking

Doyuni replied:
Oh, I missed that point. 

Thank you for telling me.
